sys/psa_crypto/psa_persistent_storage/psa_crypto_cbor_encoder.c

```diff
@@ -0,0 +1,534 @@
+    psa_get_public_key_data_from_key_slot(slot, &pubkey_data, &pubkey_data_len);
+
+    res = nanocbor_fmt_array(enc, 2); /* Contains private and public key*/
+    if (res < 0) {
+        goto error;
+    }
+
+    res = nanocbor_put_bstr(enc, privkey_data, *privkey_data_len);
+    if (res < 0) {
+        goto error;
+    }
+
+    res = nanocbor_put_bstr(enc, pubkey_data, *pubkey_data_len);
+    if (res < 0) {
+        goto error;
+    }
+
+    return PSA_SUCCESS;
+
+error:
+    DEBUG("[persistent storage] error while encoding asymmetric key: %d = %s", res,
+                                                nanocbor_error_to_string(res));
+    return nanocbor_error_to_psa_status(res);
+}
+#endif /* PSA_ASYMMETRIC_KEYPAIR_COUNT */
+
+#if PSA_SINGLE_KEY_COUNT
+/**
+ * @brief   Encode contents of a PSA key slot containing a single key
+ *
+ * @param   enc     Active nanocbor encoder
+ * @param   slot    Pointer to key slot to encode
+ * @return  psa_status_t
+ */
+static psa_status_t psa_encode_single_key(nanocbor_encoder_t *enc, psa_key_slot_t *slot)
+{
+    size_t *key_data_len;
+    uint8_t *key_data;
+    psa_get_key_data_from_key_slot(slot, &key_data, &key_data_len);
+
+    int res = nanocbor_put_bstr(enc, key_data, *key_data_len);
+    if (res < 0) {
+        DEBUG("[persistent storage] error while encoding single key: %d = %s", res,
+                                                nanocbor_error_to_string(res));
+        return nanocbor_error_to_psa_status(res);
+    }
+    return PSA_SUCCESS;
+}
+#endif /* PSA_SINGLE_KEY_COUNT */
+
+psa_status_t psa_encode_key_slot(psa_key_slot_t *slot, uint8_t *output,
+                                     size_t output_len, size_t *output_size)
+{
+    int res;
+    psa_status_t status;
+    nanocbor_encoder_t enc;
+    nanocbor_encoder_init(&enc, output, output_len);
+    *output_size = 0;
+
+    /* Key Struct */
+    res = nanocbor_fmt_array(&enc, 2);
+    if (res < 0) {
+        goto error;
+    }
+
+    status = psa_encode_key_attributes(&enc, &slot->attr);
+    if (status != PSA_SUCCESS) {
+        return status;
+    }
+
+#if PSA_PROTECTED_KEY_COUNT
+    if (psa_key_lifetime_is_external(slot->attr.lifetime)) {
+        status = psa_encode_protected_key_slot(&enc, slot);
+        if (status != PSA_SUCCESS) {
+            return status;
+        }
+        goto done;
+    }
+#endif /* PSA_PROTECTED_KEY_COUNT */
+
+#if PSA_ASYMMETRIC_KEYPAIR_COUNT
+    if (PSA_KEY_TYPE_IS_KEY_PAIR(slot->attr.type)) {
+        status = psa_encode_asymmetric_key_pair(&enc, slot);
+        if (status != PSA_SUCCESS) {
+            return status;
+        }
+        goto done;
+    }
+#endif /* PSA_ASYMMETRIC_KEYPAIR_COUNT */
+
+#if PSA_SINGLE_KEY_COUNT
+    status = psa_encode_single_key(&enc, slot);
+    if (status != PSA_SUCCESS) {
+        return status;
+    }
+    goto done;
+#endif /* PSA_SINGLE_KEY_COUNT */
+
+done:
+    *output_size = nanocbor_encoded_len(&enc);
```

Comment:
Hmmm if anyone builds PSA with a key based algorithm without at least one type of key slot, nothing would work and it would fail at key creation (maybe even during the build).
But of course I can add it just in case.

Reply:
done